### DISCUSSION TITLE
Lens model setup for DESIJ2158+0257

### DISCUSSION THREAD

**Discussion by ajshajib** [AstroBridge/BDLensing#46] [General]
@Rajit13, your model [here](https://github.com/Rajit13/BDLensing/blob/ad667711278ba70d913b713312656191a8570b4d/DESIJ2158+0257/BDLensing_DESIJ2158+0257_Trial2.ipynb) is looking great, great work! 

We would like the reduced $\chi^2$ close to 1.2 or lower. Increasing `n_max` would help alleviate the residuals on the arc, but I don't think we should go more than `n_max = 10`. Try experimenting with a few different values; the downside of higher `n_max` is that it would take more time to run the model.

You have some residual at the galaxy center and on the far right side. You can try not to join centroids and ellipticities between the mass and light and then see if the lens galaxy residual improves. If that doesn't help, masking off the center could be an option.

**Comment by Rajit13** [comment]
The current model is [this](https://github.com/Rajit13/BDLensing/blob/8bbb3b18d3477167387080bc86e83faaa799a380/DESIJ2158+0257/BDLensing_DESIJ2158+0257_Trial2.ipynb) one. It has the reduced $\chi^2$ of 1.2999. I've been still using `n_max` 10. In google collab, it takes at least 4 hours to run now. 

I've already disjoined the lens with light that's how it dropped to 1.3 and after further improvements, it got to 1.299. Now I think I'll mask the center to see what happens.

**Reply by Rajit13** [reply]
@ajshajib  vai 
After removing the center as I said I would do it got to a reduced $\chi^2=0.920911$, which is below 1. Is it normal or I just removed something important? In the last meeting, I asked you about removing the centers of a few galaxies from your papers' models.  Here is the [code](https://github.com/Rajit13/BDLensing/blob/1da19b2c19af764c412eb8042198da960f1bfdb0/DESIJ2158+0257/BDLensing_DESIJ2158+0257_Trial2.ipynb)

**Reply by ajshajib** [reply]
That value is fine. What's the radius of the central mask? Just to see what happens, try taking it no more than 0.25". Another thing you could try without taking the mask is not joining the ellipticity between the two Sersic profiles.

**Reply by Rajit13** [reply]
I used these values for the central mask `lens_center_ra+0.5, lens_center_dec-0.8, 0.7, 0.7,0` so I believe it's 0.7. So you're telling me to comment out this part `joint_lens_light_with_lens_light = [[0, 1, ['center_x', 'center_y']]]`, right? I already commented out `#joint_lens_with_light = [[0, 0, ['center_x', 'center_y', 'e1', 'e2']]]` this.

**Reply by ajshajib** [reply]
I see, `joint_lens_light_with_lens_light = [[0, 1, ['center_x', 'center_y']]]` only joins the centers of the Sersic profiles, so the ellipticities are already free between the two Sersics and no need to do anything else on that.

**Reply by Rajit13** [reply]
So what should I do now? Should I try a central mask radius 2.5''?

**Reply by ajshajib** [reply]
Yes, and the radius is 0.25".

**Comment by Rajit13** [comment]
For everyone else to see what normalized residuals plot for reduced $\chi^2=1.299..$ and $\chi^2=0.92..$ looks like for my model galaxy
![reduced](https://user-images.githubusercontent.com/81336262/235270581-0a71db87-48b8-4b62-b372-585a99f891ff.png)

**Reply by Rajit13** [reply]
**In this gif image, you can see some drastic changes made by just removing the center!**


![git](https://user-images.githubusercontent.com/81336262/235271357-ecbf6420-eeaf-4da1-b0bc-9f1a0a6f725a.gif)

**Comment by Rajit13** [comment]
@ajshajib vai, I am having difficulty running MCMC on Google Collab, it is completing 1%/10 min. I can't run the file that long. My home pc is not that strong either (corei3, 4 GB). Any suggestions or solutions regarding the issue?

**Reply by ajshajib** [reply]
Ok, I will look into other computing resources. But let's see if there's a way to make the most out of Colab. What walker ratio are you using, and how many free parameters do you have in your model?

**Reply by Rajit13** [reply]
I am using different cell to run MCMC here is the parameters
```
!pip install emcee
fitting_kwargs_list_mcmc = [['MCMC', {'n_burn': 200, 'n_run': 600, 'n_walkers':                                      
                                      200, 'sigma_scale': .1}]]

fitting_seq_mcmc = FittingSequence(kwargs_data_joint, kwargs_model, 
                                    kwargs_constraints, 
                                    kwargs_likelihood, kwargs_params)

chain_list_mcmc = fitting_seq_mcmc.fit_sequence(fitting_kwargs_list_mcmc)
kwargs_result_mcmc = fitting_seq_mcmc.best_fit()
```

**Reply by ajshajib** [reply]
I would guess the number of free parameters is 15-20, but please check. In that case, it is fine to have not more than `'n_walkers': 80`. Also, you can try to check if Colab allows multiprocessing by adding the `'threadCount'` keyword as:

```
fitting_kwargs_list_mcmc = [['MCMC', {'n_burn': 200, 'n_run': 600,  'n_walkers': 200, 'sigma_scale': .1, 'threadCount': 8}]]
```

If it works, you should get speed up by a factor of 4-6. When you want to test the speed up, it's fine to set the `'n_run'` small, like 20, which should be good enough for you to notice any difference.

`progress=True` makes things slow, so you may want to set it `False` for long runs. You can also periodically save the chain using the keywords `backend_filename` and `start_from_backend`, so that if the MCMC abruptly stops, you can resume from the last checkpoint instead of the beginning.

**Reply by Rajit13** [reply]
Will this work?
```
!pip install emcee

import os
from lenstronomy.Workflow.fitting_sequence import FittingSequence
from lenstronomy.Sampling.MCMCSampler import MCMCSampler

# Define the backend filename
backend_filename = 'mcmc_backend.h5'

# Check if a previous backend file exists
start_from_backend = os.path.isfile(backend_filename)

fitting_kwargs_list_mcmc = [['MCMC', {'n_burn': 200, 'n_run': 20, 'n_walkers': 80,
                                      'sigma_scale': .1, 'threadCount': 8,
                                      'backend_filename': backend_filename,
                                      'start_from_backend': start_from_backend}]]

fitting_seq_mcmc = FittingSequence(kwargs_data_joint, kwargs_model,
                                   kwargs_constraints,
                                   kwargs_likelihood, kwargs_params)

if start_from_backend:
    # Load the previous chain state from the backend file
    sampler = MCMCSampler(kwargs_likelihood, kwargs_params, kwargs_model,
                          fitting_seq_mcmc.kwargs_lens_eq_solver)
    sampler.load_backend(filename=backend_filename)
    fitting_seq_mcmc._sampler = sampler

chain_list_mcmc = fitting_seq_mcmc.fit_sequence(fitting_kwargs_list_mcmc)
kwargs_result_mcmc = fitting_seq_mcmc.best_fit()
```

**Reply by ajshajib** [reply]
You don't need to manually load the saved file using the backend yourself. If `'start_from_backend': True`, it will automatically load the saved file with the filename given with `'backend_filename'`, and then save the extended chain in that file as the MCMC progresses. If `'start_from_backend': False`, MCMC will start from zero and any pre-existing file with the same name as `'backend_filename'` will be overwritten with the new fresh chain. So, this part is unnecessary:

```
if start_from_backend:
    # Load the previous chain state from the backend file
    sampler = MCMCSampler(kwargs_likelihood, kwargs_params, kwargs_model,
                          fitting_seq_mcmc.kwargs_lens_eq_solver)
    sampler.load_backend(filename=backend_filename)
    fitting_seq_mcmc._sampler = sampler
```

**Reply by Rajit13** [reply]
Using the above code excluding the portion you suggested it says: `'EnsembleSampler' object has no attribute '_previous_state'`. So at first, I set it to `'start_from_backend': False`, I think I've to set it to `True` once it stops the first time.

**Reply by Rajit13** [reply]
I am supposed to see some corner graphs here right? I tried different things but couldn't figure out what to do. I checked the whole `lenstronomy` repository and found only two ipynb files on MCMC. An example/code would be helpful.

![image](https://github.com/ajshajib/BDLensing/assets/81336262/bb98a447-ba92-46e9-8dcf-d9fb62798d27)

**Reply by ajshajib** [reply]
Since we are discussing with MCMC and not particularly the lens model setup here, it would be better if if you open a new discussion thread on that topic. That would be helpful for others to find useful info in the future. Also, if any of the replies solved your original question/issue, please mark that as the answer.
